File: ruby/lrucache_test.rb
require_relative 'lrucache'
require 'test/unit'

class TestLRUCache < Test::Unit::TestCase
  def test_set
    cache = LRUCache.new(1)
    assert_equal(nil, cache.set(:test, 1))
  end

  def test_get
    cache = LRUCache.new(1)
    assert_equal(nil, cache.set(:test, 1))
    assert_equal(1, cache.get(:test))
  end

  def test_set_full_cache
    cache = LRUCache.new(1)
    assert_equal(nil, cache.set(:test, 1))
  end

  def test_set_exceptions
    assert_raise { LRUCache.new(0) }
  end

  def test_set_replace_value
    cache = LRUCache.new(1)
    cache.set(:test, 1)
    cache.set(:test, 2)

    assert_equal(2, cache.get(:test))
  end

  def test_set_expiring
    cache = LRUCache.new(2)
    cache.set(:foo, 1)
    cache.set(:bar, 2)
    cache.get(:foo)
    cache.set(:baz, 3)

    assert_equal(nil, cache.get(:bar))
    assert_equal(3, cache.get(:baz))
    assert_equal(1, cache.get(:foo))
  end

  def test_invalidate
    cache = LRUCache.new(2)
    assert_equal(nil, cache.invalidate(:foo))
    cache.set(:foo, 1)
    assert_equal(1, cache.invalidate(:foo))
  end
end

class TestLinkedList < Test::Unit::TestCase
  def test_unshift
    list = DoublyLinkedList.new
    list.unshift(1)
    assert_equal(1, list.send(:root).next.val)
    assert_equal(1, list.send(:root).prev.val)
  end

  def test_unshift_multi
    list = DoublyLinkedList.new
    list.unshift(1)
    list.unshift(2)
    assert_equal(2, list.send(:root).next.val)
    assert_equal(1, list.send(:root).prev.val)
  end

  def test_move_front
    list = DoublyLinkedList.new
    node = list.unshift(1)
    list.unshift(2)
    assert_equal(2, list.send(:root).next.val)
    assert_equal(1, list.send(:root).prev.val)

    list.move_front(node)
    assert_equal(2, list.send(:root).prev.val)
    assert_equal(1, list.send(:root).next.val)
  end
end

File: python/lrucache_test.py
import unittest
from lrucache import LRUCache 

class TestLRUCache(unittest.TestCase):
    def test_get_empty(self):
        cache = LRUCache()
        self.assertEqual(cache.get('foo'), None)

    def test_set_empty(self):
        cache = LRUCache()
        cache.set('foo', 10)
        self.assertEqual(cache.get('foo'), 10)

    def test_set_replace(self):
        cache = LRUCache()
        cache.set('foo', 10)
        cache.set('foo', 20)
        self.assertEqual(cache.get('foo'), 20)

    def test_set_full(self):
        cache = LRUCache(1)
        cache.set('foo', 10)
        cache.set('bar', 20)
        self.assertEqual(cache.get('foo'), None)
        self.assertEqual(cache.get('bar'), 20)

    def test_get_move_front(self):
        cache = LRUCache(2)
        cache.set('foo', 10)
        cache.set('bar', 20)
        cache.get('foo')
        cache.set('baz', 30)
        self.assertEqual(cache.get('foo'), 10)
        self.assertEqual(cache.get('bar'), None)
        self.assertEqual(cache.get('baz'), 30)

    def test_set_move_front(self):
        cache = LRUCache(2)
        cache.set('foo', 10)
        cache.set('bar', 20)
        cache.set('foo', 40)
        cache.set('baz', 30)
        self.assertEqual(cache.get('foo'), 40)
        self.assertEqual(cache.get('bar'), None)
        self.assertEqual(cache.get('baz'), 30)

    def test_invalidate_empty(self):
        cache = LRUCache(2)
        self.assertEqual(cache.invalidate('foo'), None)

    def test_invalidate(self):
        cache = LRUCache(2)
        cache.set('foo', 10)
        self.assertEqual(cache.get('foo'), 10)
        self.assertEqual(cache.invalidate('foo'), 10)
        self.assertEqual(cache.get('foo'), None)
        

if __name__ == '__main__':
    unittest.main()
